torchaudio/torch_sox.cpp

```diff
@@ -0,0 +1,117 @@
+#include <torch/torch.h>
+
+#include <sox.h>
+
+#include <algorithm>
+#include <cstdint>
+#include <stdexcept>
+#include <vector>
+
+namespace torch { namespace audio {
+int read_audio_file(const std::string& file_name, at::Tensor output) {
+  if (sox_init() != SOX_SUCCESS) {
+    throw std::runtime_error("Error initializing sox library");
+  }
+  sox_format_t* fd = sox_open_read(
+      file_name.c_str(),
+      /*signal=*/nullptr,
+      /*encoding=*/nullptr,
+      /*filetype=*/nullptr);
+  if (fd == nullptr) {
+    throw std::runtime_error("Error opening audio file");
+  }
+
+  const int64_t number_of_channels = fd->signal.channels;
+  const int sample_rate = fd->signal.rate;
+  const int64_t buffer_length = fd->signal.length;
+  if (buffer_length == 0) {
+    throw std::runtime_error("Error reading audio file: unknown length");
+  }
+
+  std::vector<sox_sample_t> buffer(buffer_length);
+  const int64_t samples_read = sox_read(fd, buffer.data(), buffer_length);
+  if (samples_read == 0) {
+    throw std::runtime_error(
+        "Error reading audio file: empty file or read failed in sox_read");
+  }
+
+  output.resize_({samples_read / number_of_channels, number_of_channels});
+  output = output.contiguous();
+
+  AT_DISPATCH_ALL_TYPES(output.type(), "read_audio_buffer", [&] {
+    auto* data = output.data<scalar_t>();
+    std::copy(buffer.begin(), buffer.begin() + samples_read, data);
+  });
+
+  sox_quit();
+
+  return sample_rate;
+}
+
+void write_audio_file(
+    const std::string& file_name,
+    at::Tensor tensor,
+    const std::string& extension,
+    int sample_rate) {
+  if (sox_init() != SOX_SUCCESS) {
+    throw std::runtime_error("Error initializing sox library");
+  }
+
+  if (!tensor.is_contiguous()) {
+    throw std::runtime_error(
+        "Error writing audio file: input tensor must be contiguous");
+  }
```

Comment:
This is missing a test that the input tensor is 2D. (The original code was missing this too, but it didn't use `numel` to compute length, so you wouldn't ever read out of bounds. This code can easily read out of bounds if you have a multi-dim tensor.)